src/os_auth/auth.c

```diff
@@ -164,62 +166,88 @@ w_err_t w_auth_parse_data(const char* buf, char *response,const char *authpass,
         }
 
     }
-    else if(!config.flags.use_source_ip) {
-        // use_source-ip = 0 and no -I argument in agent
-        snprintf(ip, IPSIZE, "any");
+    else{
+        buf--;
+        if(!config.flags.use_source_ip) {
+            // use_source-ip = 0 and no -I argument in agent
+            snprintf(ip, IPSIZE, "any");
+        }
+    }
+
+    /* Check for key hash when the agent already has one*/
+    char key_hash_token[2] = "K:";
+    if(strncmp(++buf, key_hash_token, 2)==0)
+    {
+        os_calloc(1, sizeof(os_sha1), *key_hash);
+        char format[15] = {0};
+        sprintf(format, " K:\'%%%ld[^\']\"", sizeof(os_sha1));
+        sscanf(buf, format, *key_hash);
     }
-    // else -> agent IP is already on ip
 
     return OS_SUCCESS;
 }
 
-w_err_t w_auth_validate_data (char *response, const char *ip, const char *agentname, const char *groups){
-    /* Validate the group(s) name(s) */
-    int index = 0;
-    char *id_exist = NULL;
+w_err_t w_auth_replace_agent(keyentry *key,
+                             const char *key_hash){
+    /* Check if the agent antiquity complies with the configuration to be removed*/
     double antiquity = 0;
+    if (!config.flags.force_insert ||
+       (antiquity = OS_AgentAntiquity(key->name, key->ip->ip), antiquity > 0 && antiquity < config.force_time)) {
+        minfo("Agent '%s' doesn´t complies with the antiquity to be removed.", key->id);
+        return OS_INVALID;
+    }
+
+    /* Check if the agent key is the same than the existent in the manager */
+    if (key_hash) {
+        os_sha1 manager_key_hash;
+        w_auth_hash_key(key, manager_key_hash);
+        if (!strcmp(manager_key_hash, key_hash)) {
+            minfo("Key identified with hash '%s' already exists on the manager.", key_hash);
```

Comment:
We can discuss with the team if it's a good idea to log the hash of the key.

Reply:
I think It doesn't have much value. Probably would be better to log something like...

`Agent '%s' key already exists on the manager.`